Sources/Operators/ConcatMap.swift

```diff
@@ -0,0 +1,163 @@
+//
+//  ConcatMap.swift
+//  CombineExt
+//
+//  Created by Daniel Peter on 22/11/2020.
+//  Copyright © 2020 Combine Community. All rights reserved.
+//
+
+#if canImport(Combine)
+import Combine
+import class Foundation.NSRecursiveLock
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public extension Publisher {
+    /// Transforms an output value into a new publisher, and flattens the stream of events from these multiple upstream publishers to appear as if they were coming from a single stream of events.
+    ///
+    /// Mapping to a new publisher will keep the subscription to the previous one alive until it completes and only then subscribe to the new one. This also means that all values sent by the new publisher are not forwarded as long as the previous one hasn't completed.
```

Comment:
```suggestion
    /// Mapping to a new publisher will keep the subscription to the previous one alive until it completes and only then subscribe to the new one. This also means that all values sent by the new publisher are not forwarded as long as the previous one hasn't completed.
```

> This also means that all values sent by the new publisher are not forwarded as long as the previous one hasn't completed.

Not sure about this - why would it emit at all if it's only subscribed to after the completion of the previous one?

Reply:
The inner publisher could potentially emit before being subscribed to (Like a non-deferred future, i.e. hot observable)?


Example:
```swift
let passthrough = PassthroughSubject<Int, Never>()

let mapped = passthrough.map { value in
    Future<Int, Never> { completion in
        completion(.success(value))
    }
    .delay(for: .seconds(1), scheduler: DispatchQueue.main)
  }
  .compactMap { $0 }

  let cancellable = mapped.sink(receiveValue: { value in
    print(value)
  })

  passthrough.send(1)
  passthrough.send(2)     
```

In this case, concatMap does not **buffer** 2 as the Publisher 1 hadn't completed when 2 was put onto the stream.

See `test_ignores_values_of_subsequent_while_previous_hasNot_completed` in `CompactMapTests.swift`

Reply:
In Rx, the concatMap assumes that inner Observables are cold and that the closure creates the Observable that is being subscribed to. I think it's correct to make that assumption here as well.